### PR TITLE
Updated MWS implementation

### PR DESCRIPTION
**Pull request by abailoni**
Summary of changes:

- More efficient implementation of the mutex storage (only keep max one constraint between two clusters). In all my tests this implementation is quite faster
- Updated `compute_mws_segmentation_impl`: it now requires a boolean array specifying the mutex edges (instead of using the convention of the `number_of_attractive_channels`). The python function `compute_mws_segmentation` keeps the same signature. And I added a new python function  `compute_mws_segmentation_from_signed_affinities` that considers as Mutex edges all those with negative weights

I made these changes a while ago, so I already had a chance to properly test them. At the moment the GitHub workflow fails because of some ImportErrors during tests. 

Apart from that, on master for me the tests fail because of some other import error and a Segmentation fault. However, I tested `test_mutex_watershed.py` on this branch (which involves the updated functions) and that works for me.

### CODE REVIEW

**Review by constantinpape — COMMENTED** [review]
I only gave the changes a quick look now, and given that this is changing the internal implementation quite significantly we will need to:
- add tests that check the correctness of the result w.r.t. previously computed results
- add a benchmark that actually checks the performance 

I can add both of this on master so you can merge it here and evaluate the benchmark.

Regarding the change to a set datastructure for the mutex storage: I think this makes a lot of sense; we also had a version with this but never fully debugged it. But I would think that an unordered_set should be even more performant. Again, this should be benchmarked.

Regarding the test failures: I don't quite understand what's going on there, I will check if the CI also fails on master.

**Comment by constantinpape** [comment]
I added a reference test on master, started to work on benchmarks in #18 and will also use this PR to debug tests.

File: include/affogato/segmentation/mutex_watershed.hxx [hunk 1]
```diff
@@ -115,7 +102,8 @@ namespace segmentation {
         });
 
         // data-structure storing mutex edges
-        typedef std::vector<std::vector<uint64_t>> MutexStorage;
+        typedef boost::container::flat_set<uint64_t> SetType;
```
[the remark this answers is not in the document]
Comment (abailoni) [reply include/affogato/segmentation/mutex_watershed.hxx:105]: I agree, adding tests and benchmarks in the master makes total sense. I believe I did it for both modifications when I implemented them, but even better if we have it in the tests.

Regarding the set data-structure, I thought that finding constraints would have been faster, but it indeed makes sense to benchmark several options and see.

**Comment by constantinpape** [comment]
@abailoni I have  updated master now, so that all tests run (and added the tests to compare the mws results to a reference output).
I have also added a benchmark script: https://github.com/constantinpape/affogato/blob/master/script/benchmark.py. Please run it for
- the current implementation using `std::vector`
- using a `boost::flat_set`
- using `std::unordered_set` (I think this should not need any other changes compared to using the flat set).

I will add a plot script later to compare the benchmark results.

**Comment by constantinpape** [comment]
I have added the plot script here: https://github.com/constantinpape/affogato/blob/master/script/plot_benchmark.py

**Comment by abailoni** [comment]
`boost::flat_set` wins by a little margin over `std::unordered_set`. But the improvement over `std::vector` is consistent.

benchmark

**Comment by constantinpape** [comment]
> `boost::flat_set` wins by a little margin over `std::unordered_set`. But the improvement over `std::vector` is consistent.

Awesome, then `flat_set` it is :).

File: src/python/module/affogato/segmentation/mws.py [hunk 2]
```diff
@@ -4,8 +4,8 @@
 from ..affinities import compute_affinities
 
 
-def get_valid_edges(shape, offsets, number_of_attractive_channels,
-                    strides, randomize_strides, mask=None):
+def get_valid_edges(shape, offsets, number_of_attractive_channels=None,
+                    strides=None, randomize_strides=False, node_mask=None, edge_mask=None):
```
Comment (constantinpape) [inline src/python/module/affogato/segmentation/mws.py:8]: What happens if we have both a `node_mask` and an `edge_mask`, does it result in a well defined state?
If not we should raise some exception int hat case.

**Review by constantinpape — COMMENTED** [review]
Looks good now, just one  last question about the node mask / edge mask.

File: src/python/module/affogato/segmentation/mws.py (hunk 2 above)
Comment (abailoni) [reply src/python/module/affogato/segmentation/mws.py:8]: At the moment, when a `node_mask` is passed, all edges connected to a background node are masked. Then, if an `edge_mask` is also passed and there are extra edges that were not already masked, they are also masked.

**Pull request merged**